tee/src/lib.rs

```diff
@@ -0,0 +1,669 @@
+		/// Unknown ClusterId
+		UnknownClusterId,
+		/// Account does not associated with an enclave
+		NotEnclaveOwner,
+		/// Enclave address registered to an account
+		EnclaveAddressAlreadyRegisteredtoTheAccount,
+		/// Enclave URI is short
+		UriTooShort,
+		/// Enclave URI is long
+		UriTooLong,
+		/// Maximum enclaves reached
+		EnclaveIdOverflow,
+		/// Maximum clusters reached
+		ClusterIdOverflow,
+		/// Cluster is already full, cannot assign any enclaves
+		ClusterIsAlreadyFull,
+		/// Enclave already assigned to a cluster
+		EnclaveAlreadyAssigned,
+		/// Enclave not assigned to a cluster
+		EnclaveNotAssigned,
+		/// Enclave does not exists
+		EnclaveDoesNotExists,
+		/// Internal logical error
+		InternalLogicalError,
+		/// Unknown enclave operator account
+		UnknownEnclaveOperatorAccount,
+		/// Error Adding To Queue
+		ErrorAddingToQueue,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::register_enclave())]
+		pub fn register_enclave(
+			origin: OriginFor<T>,
+			enclave_address: Vec<u8>,
+			api_uri: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let account = ensure_signed(origin)?;
+
+			ensure!(api_uri.len() < T::MaxUriLen::get().into(), Error::<T>::UriTooLong);
+			ensure!(api_uri.len() > T::MinUriLen::get().into(), Error::<T>::UriTooShort);
+
+			let result = AccountEnclaveId::<T>::get(&account)
+				.and_then(|enclave_id| EnclaveData::<T>::get(enclave_id).map(|enc| enc));
+
+			if result.is_some() {
+				let enclave = result.ok_or(Error::<T>::EnclaveDoesNotExists)?;
+				ensure!(
+					enclave.enclave_address != enclave_address,
+					Error::<T>::EnclaveAddressAlreadyRegisteredtoTheAccount
+				)
+			}
+
+			let (enclave_id, new_id) = Self::new_enclave_id()?;
+			// Needs to have enough money
+			let imbalance = T::Currency::withdraw(
+				&account,
+				T::EnclaveFee::get(),
+				WithdrawReasons::FEE,
+				KeepAlive,
+			)?;
+			T::FeesCollector::on_unbalanced(imbalance);
+
+			let enclave = Enclave::new(api_uri.clone(), enclave_address);
+
+			AccountEnclaveId::<T>::insert(account.clone(), enclave_id);
+			EnclaveData::<T>::insert(enclave_id, enclave);
+			EnclaveIdGenerator::<T>::put(new_id);
+
+			let reg_enclaves: BoundedVec<EnclaveId, T::MaxRegisteredEnclaves> =
+				<EnclaveRegistrationList<T>>::get();
+
+			let reg_enclaves = reg_enclaves
+				.try_mutate(|v| v.push(enclave_id))
+				.ok_or(Error::<T>::ErrorAddingToQueue)?;
+			<EnclaveRegistrationList<T>>::put(reg_enclaves);
+
+			Self::deposit_event(Event::EnclaveAdded { account, api_uri, enclave_id });
+			Ok(().into())
+		}
+
+		/// Removes an enclave from the system
+		/// Origin- operator account address
+		/// If the enclave is assigned, it will be placed in queue for tech committee approval
+		/// If enclave is not already assigned, he can exit without permission.
+		#[pallet::weight(T::WeightInfo::unregister_enclave())]
+		pub fn unregister_enclave(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
```

Comment:
This function is only used by the enclave operator, not root, so only the basic checks are needed. Also, it's not the enclave address, but the enclave operator's address. I don't see the enclave operator to enclave Id mapping only AccountEnclaveId.

Reply:
Can we discuss this ? as per the last discussion this has to be call by the root

Reply:
The mapping  happens here

````
	#[pallet::storage]
	#[pallet::getter(fn enclave_index)]
	pub type AccountEnclaveId<T: Config> =
		StorageMap<_, Blake2_128Concat, T::AccountId, EnclaveId, OptionQuery>;
```
We have removed operator registration. Please provide the correct flow 
@Leouarz